Rearranging funcs and methods

Dann38 opened the pull request.

Dann38 committed:
move function get_rid_spaces and base_line_image from utils in ClasterizationBoldClassifier

Dann38 committed:
move method preprocessing in ClasterizationBoldClassifie
add self.binarizer

From hist, mean and ps _bold_classifier

Dann38 committed:
fix import

Dann38 committed:
add evaluate.py

main evaluate.py
move llist2vector and vector2llist in utils.py (bold_classifier)

Dann38 committed:
refactor code project

refactor variables name
change 1 -> BOLD, 0 -> REGULAR
change import order of libraries

Dann38 committed:
refactor code project

refactor variables name
change 1 -> BOLD, 0 -> REGULAR
change import order of libraries

Dann38 committed:
fix f1

Dann38 committed:
change output get_bboxes_and_style

Dann38 committed:
add examples

Dann38 committed:
add types_font

Dann38 committed:
update README.md

Dann38 committed:
create private method

Dann38 committed:
change name class ClusterizationBoldClassifier

dronperminov commented on examples/save_dir_mark_image.py:
```diff
@@ -0,0 +1,22 @@
+
+pages = reader.get_array_pages(name_data)
+drawer = Drawer()
+for i in range(len(pages)):
```
не стоит использовать конструкцию `for i in range(len(pages))`, если используются элементы коллекции
лучше использовать запись `for i, page in enumerate(pages)`

Dann38 replied:
done

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,102 @@
+        print("*" * 15, name_dataset, "*" * 15)
+        for classifier in classifier_list:
+            cl = classifier()
+            cl.clusterizater.significance_level = 0.5
```
далеко не у каждого классификатора может быть кластеризатор и более того, менять что-то после создания объекта очень плохой тон

Dann38 replied:
ок

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,102 @@
+
+        print("*" * 15, name_dataset, "*" * 15)
+        for classifier in classifier_list:
+            cl = classifier()
```
Не стоит использовать список наименований класса, лучше в `main` честно создать список экземпляров классификаторов

Dann38 replied:
ok

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,102 @@
+        print(f"{kei + ':':{12}} {evaluate[kei]:.2f}")
+
+
+if __name__ == "__main__":
```
Стоит использовать функцию `main`, которую вызывать в данном месте, поскольку в таком случае все объявленные переменные торчат в глобальную область видимости

Dann38 replied:
ok

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,102 @@
+
+
+def print_evaluate(evaluate: Dict) -> None:
+    print("-" * 10, classifier.__name__, "-" * 10)
```
использовать слота `__name__` не самый лучший выбор. Может лучше вместо списка классификаторов тогда сделать словарь примерно такого вида:
```
classifiers = {
    "PsBoldClassifier": PsBoldClassifier(),
    "MeanBoldClassifier": MeanBoldClassifier(),
    "HistBoldClassifier": HistBoldClassifier()
}
```

И затем при тестировании идти по ним так:
`for classifier_name, classifier in classifiers.items():`

Dann38 replied:
ok

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,102 @@
+    return evaluate_sum
+
+
+def evaluate_llist(llist: List[List[float]], llist_true: List[List[float]]) -> Dict:
```
быть может использовать стандартные функции из `sklearn.metrics` вместо самостоятельного подсчёта?
можно поглядеть примеры [здесь](https://vitalflux.com/accuracy-precision-recall-f1-score-python-example/)

Dann38 replied:
ок

dronperminov commented on bold_classifier/clusterization_bold_classifier/hist_bold_classifier/hist_bold_classifier.py:
```diff
@@ -0,0 +1,17 @@
+        img = self._get_rid_spaces(img)
+        if len(img) == 0:
+            return 1
+        if img.shape[1] == 0:
```
есть ли смысл в отдельном `if`, если результат в обоих случаях одинаковый?

Dann38 replied:
ok

dronperminov commented on dataset_reader/dataset_reader.py:
```diff
@@ -13,12 +13,12 @@ def get_array_pages(self, path_dir: str) -> List[Page]:
         return pages
 
-    def is_page(self, path_dir: str, name: str) -> bool:
+    def _is_page(self, path_dir: str, name: str) -> bool:
```
точно ли этот (и последующие аналогичные методы) должны быть `protected`? Если наследники не планируются, то лучше делать их `private`

Dann38 replied:
ok

dronperminov commented on dataset_reader/dataset_reader.py:
```diff
@@ -13,12 +13,12 @@ def get_array_pages(self, path_dir: str) -> List[Page]:
         files = os.listdir(path_dir)
         pages = []
```
Быть может сделать покороче через списковое включение?
```
pages = [self.get_page(path_dir, name_file) for name_file in files if self._is_page(path_dir, name_file)]
```

Dann38 replied:
ok

dronperminov commented on dataset_reader/dataset_reader.py:
```diff
@@ -36,24 +36,23 @@ def get_page(self, path_dir: str, name: str) -> Page:
 
-    def get_image(self, path_dir: str, name_image_file: str) -> np.ndarray:
+    def _get_image(self, path_dir: str, name_image_file: str) -> np.ndarray:
         path_image = os.path.join(path_dir, name_image_file)
```
а точно просто `cv2.imread` недостаточно? Для чего читать изображение бинарно?

Dann38 replied:
На русском языке путь не читается (по крайней мере в Windows)
https://answers.opencv.org/question/205345/imread-and-russian-language-path-to-img/

dronperminov commented on clusterizater/clusterizater.py:
```diff
@@ -1,60 +1,8 @@
 import numpy as np
 
 
 class BaseClusterizater(ABC):
```
`BaseClusterizer`

Dann38 replied:
ok

dronperminov commented on clusterizater/clusterizater.py:
```diff
@@ -1,60 +1,8 @@
 class BaseClusterizater(ABC):
     @abstractmethod
-    def clusterization(self, X: np.ndarray) -> np.ndarray:
+    def clusterization(self, x: np.ndarray) -> np.ndarray:
```
в подобных классах лучше называть методы не просто действием, а призывом к действию
Кластеризатор должен кластеризовать: `clusterize`

Dann38 replied:
ok

dronperminov commented on bold_classifier/clusterization_bold_classifier/clusterization_bold_classifier.py:
```diff
@@ -0,0 +1,85 @@
+
+
+class ClusterizationBoldClassifier(BaseBoldClassifier):
+    def __init__(self, binarizer=ValleyEmphasisBinarizer, clusterizater=BoldSpectralClusterizater):
```
не стоит передавать в качестве аргументов имена классов, плохой тон :)
есть ощущение, что в итоге были выбраны именно эти кластеризаторы как самые лучшие, в противном случае, хотелось бы видеть сравнение их в `evaluate.py`.
Если же моя теория верна, то мб и нет смысла их плодить?

Dann38 replied:
Не знаю какую кластеризацию выбрать, сделал общую проверку гипотезы об однородности классов (_is_homogeneous) и Bold2MeanClusterizer стал работать в некоторых случаях лучше, чем BoldSpectralClusterizer

Dann38 committed:
change "for" in save_dir_mark_image.py
```
for i in range(len(pages)):
num_page = i
f'{i}.jpeg'
```->
```
for page in pages:
page.name
```

Dann38 committed:
add main-function

Dann38 committed:
change my metrics to metrics from sklearn

Dann38 committed:
change protected method to private

Dann38 committed:
change get_array_pages

Dann38 committed:
clusterizATer -> clusterizer

Dann38 committed:
clusterizATer -> clusterizer

Dann38 committed:
clusterizATer -> clusterizer

Dann38 committed:
add clasterizers evaluation

Dann38 committed:
add clasterizers evaluation

Dann38 committed:
add function _is_homogeneous

Dann38 committed:
fix two "if len(img)"

dronperminov commented on bold_classifier/clusterization_bold_classifier/clusterization_bold_classifier.py:
```diff
@@ -0,0 +1,86 @@
+        return self.binarizer.binarize(image)
+
+    def __get_evaluation_bboxes(self, image: np.ndarray,
+                               bboxes: List[List[BBox]]) -> List[List[float]]:
```
кажется, лишний перенос строки влез в код :)

Dann38 replied:
ok

dronperminov commented on bold_classifier/clusterization_bold_classifier/clusterization_bold_classifier.py:
```diff
@@ -0,0 +1,86 @@
+            for bbox in line:
+                image_bbox = image[bbox.y_top_left:bbox.y_bottom_right,
+                                   bbox.x_top_left:bbox.x_bottom_right]
+                evaluation_bbox = self.evaluation_method(image_bbox)
```
вроде как  `evaluation_method` возвращает `float`, а не `bbox`, наименования **слегка** путают

Dann38 replied:
А если сделать наоборот bbox_evaluation  (оценка bbox), это будет лучше читаться?

dronperminov commented on bold_classifier/clusterization_bold_classifier/clusterization_bold_classifier.py:
```diff
@@ -0,0 +1,86 @@
+        return evaluation_bboxes
+
+    @abstractmethod
+    def evaluation_method(self, image: np.ndarray) -> float:
```
быть может дать более осмысленное название? То, что это метод, следует из описания его как метода класса
Судя по всему он оценивает некоторую среднюю яркость изображения? Я не совсем понимаю, что именно происходит в них, но, кажется, хочется более понятное название (но это ИМХО)

Dann38 replied:
Могу предложить "evaluation_one_bbox_image" (оценка изображения одного bbox)

dronperminov commented on bold_classifier/clusterization_bold_classifier/clusterization_bold_classifier.py:
```diff
@@ -0,0 +1,86 @@
+        max2 = 0
+        argmax1 = 0
+        argmax2 = 0
+        for i in range(len(delta_mean)):
```
вчера я писал про то, что `for i in range(len(collection))` очень нехорошая практика :)
Что на счёт такого варианта?

```
for i, delta in enumerate(delta_mean):
    if delta <= max2:
        continue

    if delta > max1:
        max2 = max1
        argmax2 = argmax1
        max1 = delta
        argmax1 = i
    else:
        max2 = delta
        argmax2 = i
```

Dann38 replied:
ok

dronperminov commented on clusterizer/clusterizater.py:
```diff
@@ -0,0 +1,25 @@
+    def clusterize(self, x: np.ndarray) -> np.ndarray:
+        pass
+
+    def _is_homogeneous(self, x, x_clust0, x_clust1):
```
не хватает typing'ов

dronperminov commented on clusterizer/spectr_clusterizer/bold_spectr_clusterizer.py:
```diff
@@ -0,0 +1,34 @@
+
+class BoldSpectralClusterizer(BaseClusterizer):
+    def clusterize(self, x: np.ndarray) -> np.ndarray:
+        nearby_x = x.copy()
```
Кажется, подготовка данных перед собственно кластеризацией происходит одна и та же, может быть её тогда вынести в базовый класс?

dronperminov commented on clusterizer/spectr_clusterizer/bold_spectr_clusterizer.py:
```diff
@@ -0,0 +1,34 @@
+        spectr.fit(x_vec)
+        x_clust = spectr.labels_
+
+        x_clust0 = x[x_clust == 0]
```
аналогично от этой части код дублируется, может тоже вынести в базовый?
и тогда назвать не `BaseClusterizer`, а `BaseBoldClusterizer`

dronperminov commented on dataset_reader/dataset_reader.py:
```diff
@@ -36,24 +32,23 @@ def get_page(self, path_dir: str, name: str) -> Page:
         return page
 
-    def get_image(self, path_dir: str, name_image_file: str) -> np.ndarray:
+    def _get_image(self, path_dir: str, name_image_file: str) -> np.ndarray:
```
он точно должен быть protected?

Dann38 replied:
нет, я исправил

dronperminov commented on dataset_reader/dataset_reader.py:
```diff
@@ -36,24 +32,23 @@ def get_page(self, path_dir: str, name: str) -> Page:
-    def get_bboxes_and_style(self, path_dir: str,
-                             name_pkl_file: str) -> (List[List[int]], List[List[int]]):
+    def __get_bboxes_and_style(self, path_dir: str,
+                              name_pkl_file: str) -> (List[List[BBox]], List[List[float]]):
```
лишний перенос строки + для кортежа есть тип `Tuple` в `typing`

Dann38 replied:
ok

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,101 @@
+
+
+if __name__ == "__main__":
+    list_name_dataset = ["ВКР", "ГОСТ", "Геометрия"]
```
мне казалось, я вчера писал про `def main()`, чтобы не создавать глобальных переменных вовсе :)

Dann38 replied:
Я сначала не понял и вынес только сам процесс в функцию `main`. Сейчас вообще там все поменял :) 
Для одних документов одна кластеризация работает лучше чем другая. 
Когда я их добавлял, оказалось, что иногда  оценки однородные (весь текст обычный). Нашел **критерий**, позволяющее принимать решение между выбором одного класса (обычный текст) или двумя (жирный и обычный текст). 

Сегодня изменил кластеризацию, которая раньше была по порогу, сейчас выбирается порог исходя из лучшего значения этого **критерия** (чем он меньше, тем точнее выбрано разбиение на 2 класса). В итоге сейчас нет лучшей кластеризации, где-то одна лучше, где-то другая. 

Нужно отметить, что в кластеризации есть параметр уровня значимости, чем он ниже, тем выше точность распознания жирного (полнота соответственно падает).

dronperminov commented on dataset_reader/dataset_reader.py:
```diff
@@ -11,14 +11,10 @@
 class Reader:
     def get_array_pages(self, path_dir: str) -> List[Page]:
```
есть ли смысл назвать `get_array_pages`?
Быть может лучше `get_pages` или вовсе `read` (он же всё-таки ридер)?

Dann38 replied:
ок

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,101 @@
+    }
+    check_classifier_dataset(classifiers, list_name_dataset)
+    print("=" * 60)
+    classifiers = {
```
Почему бы сразу не одним словарём?
```
classifiers = {
    "PsBoldClassifier": PsBoldClassifier(),
    "MeanBoldClassifier": MeanBoldClassifier(),
    "HistBoldClassifier": HistBoldClassifier(),
    "PsBoldClassifier with SpectralClusterizer": PsBoldClassifier(clusterizer=BoldSpectralClusterizer()),
    "PsBoldClassifier with MeanClusterizer": PsBoldClassifier(clusterizer=Bold2MeanClusterizer()),
    "PsBoldClassifier with FixedThresholdClusterizer": PsBoldClassifier(clusterizer=BoldFixedThresholdClusterizer())
}
```

Dann38 replied:
Как оказалось лучше себя проявил (с точки зрения F1) PsBoldClassifier, выбор кластеризации как-то зависит от данных (не могу выделить лучшую, Bold2MeanClusterizer более простая, от того и думаю что надежнее) 

Прогонку изменил как уже сказал в предыдущем комментарии

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,101 @@
+
+from clusterizer import *
+
+path_dir_dataset = os.path.join(os.getcwd(), "dataset")
```
лушче не заводить глобальных переменных, не являющихся константами, а переносить это в `main`

Dann38 replied:
ок

dronperminov commented on examples/evaluate.py:
```diff
@@ -0,0 +1,101 @@
+import os
```
быть может всё-таки не в примеры этот файл, а как был в корень?

Dann38 replied:
Тут же вроде главные сами классы, а это пример использования. Можно сказать, что папка example это пример "output" работы класса, а папки dataset, dataset_reader это пример "input"

Dann38 committed:
fix "line break"

Dann38 committed:
rename in ClusterizationBoldClassifier

Dann38 committed:
change evaluate

Commit:
move types_font.py to bold_classifier

Commit:
change "for range" to "for enumerate"

Commit:
add BaseBoldClusterizer

Commit:
update

Dann38 committed:
add search the best (classifier, clusterizer)

Dann38 committed:
improve BoldFixedThresholdClusterizer

Dann38 committed:
rename HistBoldClassifier to MedianBoldClassifier

Dann38 committed:
improve the quality of the code

Dann38 committed:
private __get_image

Dann38 committed:
add typing (Tuple)

dronperminov commented on bold_classifier/clusterization_bold_classifier/clusterization_bold_classifier.py:
```diff
@@ -0,0 +1,86 @@
+
+
+class ClusterizationBoldClassifier(BaseBoldClassifier):
+    def __init__(self, clusterizer: BaseClusterizer = Bold2MeanClusterizer()):
```
давай лучше уберём значене по умолчанию, PEP рекомендует в качестве параметров по умолчанию ставить только базовые типы вроде `int`, `float`, `str`

Commit:
default type None

Dann38 committed:
fix __get_f_criterion_homogeneous

The original source has been found. There is a typo in the article
